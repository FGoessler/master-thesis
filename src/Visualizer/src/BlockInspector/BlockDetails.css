.json {
    max-width: 100%;
    word-wrap: break-word;
    word-break: break-all;
    white-space: normal;
    font-family: monospace;
    background-color: aliceblue;
}