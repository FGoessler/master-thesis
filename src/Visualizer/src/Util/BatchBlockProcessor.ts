type Placeholder = null;

// TODO: unit test this!

export default class BatchBlockProcessor<T> {

  private dataFetchingAction: (blockNumber: number) => Promise<T>;
  private dataUpdateCallback?: (datapoints: T[], labels: string[]) => void;

  constructor(
    dataFetchingAction: (blockNumber: number) => Promise<T>,
    dataUpdateCallback?: (datapoints: T[], labels: string[]) => void
  ) {
    this.dataFetchingAction = dataFetchingAction;
    this.dataUpdateCallback = dataUpdateCallback;
  }

  public async startLoading(
    fromBlockNumber: number,
    count: number,
    batchSize: number = 100
  ) {
    let datapoints = [] as T[];
    let labels = [] as string[];
    let nextBlockNumber = fromBlockNumber;

    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      let data = await this.loadBatch(nextBlockNumber, batchSize);
      datapoints = datapoints.concat(data.datapoints);
      labels = labels.concat(data.labels);
      this.sendUpdate(datapoints, labels);
      nextBlockNumber = nextBlockNumber - batchSize;
    }

    return {datapoints, labels};
  }

  private async loadBatch(fromBlockNumber: number, count: number) {
    let datapoints = [] as (T | Placeholder)[];
    let labels = [] as string[];
    let promises = [];
    let nextBlockNumber = fromBlockNumber;
    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      promises.push(this.loadDataAt(nextBlockNumber, datapoints, labels));
      nextBlockNumber = nextBlockNumber - 1;
    }
    await Promise.all(promises);
    return {datapoints: datapoints as T[], labels};
  }

  private async loadDataAt(
    blockNumber: number,
    targetDataArray: (T | Placeholder)[], targetLabelArray: string[]
  ) {
    const targetIndex = targetDataArray.length;

    targetDataArray.push(null);   // push placeholder value
    targetLabelArray.push(`${blockNumber}`);

    targetDataArray[targetIndex] = await this.dataFetchingAction(blockNumber);
  }

  private sendUpdate(datapoints: T[], labels: string[]) {
    if (this.dataUpdateCallback) {
      this.dataUpdateCallback(
        datapoints.slice(0, datapoints.length),
        labels.slice(0, datapoints.length),
      );
    }
  }
}
