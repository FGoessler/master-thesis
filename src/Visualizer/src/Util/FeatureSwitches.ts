
let featureSwitches = {
  SHOW_ADDRESS_BALANCE_CHART_DATA: false
};

export default class FeatureSwitchService {
  public static setFeatureSwitch(name: string, val: boolean) {
    featureSwitches[name] = val;
  }

  public static getFeatureSwitch(name: string): boolean {
    return !!featureSwitches[name];
  }
}
