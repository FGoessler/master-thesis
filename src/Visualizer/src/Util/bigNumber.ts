const BigNumberLib = require("bignumber.js");

const unitMap = {
  "wei":          "1",
  "kwei":         "1000",
  "ada":          "1000",
  "femtoether":   "1000",
  "mwei":         "1000000",
  "babbage":      "1000000",
  "picoether":    "1000000",
  "gwei":         "1000000000",
  "shannon":      "1000000000",
  "nanoether":    "1000000000",
  "nano":         "1000000000",
  "szabo":        "1000000000000",
  "microether":   "1000000000000",
  "micro":        "1000000000000",
  "finney":       "1000000000000000",
  "milliether":   "1000000000000000",
  "milli":        "1000000000000000",
  "ether":        "1000000000000000000",
  "kether":       "1000000000000000000000",
  "grand":        "1000000000000000000000",
  "einstein":     "1000000000000000000000",
  "mether":       "1000000000000000000000000",
  "gether":       "1000000000000000000000000000",
  "tether":       "1000000000000000000000000000000"
};

export interface BigNumber {
  toNumber(): number;
  toString(base: number): string;
  times(num: number | BigNumber): BigNumber;
  dividedBy(num: number | BigNumber): BigNumber;
}

export class BigNumberHelper {

  static toBigNumber(val: number | string | BigNumber = 0): BigNumber {
    if (val instanceof BigNumberLib) {
      return val as BigNumber;
    }

    if (typeof val === "string" && (val.indexOf("0x") === 0 || val.indexOf("-0x") === 0)) {
      return new BigNumberLib(val.replace("0x", ""), 16);
    }

    return new BigNumberLib((<number> val).toString(), 10);
  }

  static isBigNumber(object: {}) {
    return object instanceof BigNumberLib;
  }

  static toWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).times(BigNumberHelper.getValueOfUnit(unit));
  }

  static fromWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).dividedBy(BigNumberHelper.getValueOfUnit(unit));
  }

  private static getValueOfUnit(unit: string) {
    unit = unit ? unit.toLowerCase() : "ether";
    let unitValue = unitMap[unit];
    if (!unitValue) {
      throw new Error("Unit does not exist. Possible units: " + JSON.stringify(unitMap));
    }
    return new BigNumberLib(unitValue, 10);
  }
}
