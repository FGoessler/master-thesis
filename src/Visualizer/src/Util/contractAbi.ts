export const abi = [{
  "constant": false,
  "inputs": [],
  "name": "finalizeCloseThroughServiceConsumer",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminated",
  "outputs": [{"name": "", "type": "bool"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceConsumer",
  "outputs": [{"name": "", "type": "address"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "tokenRedemptionTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [{"name": "numRequests", "type": "uint256"}, {"name": "hash", "type": "bytes32"}, {
    "name": "v",
    "type": "uint8"
  }, {"name": "r", "type": "bytes32"}, {"name": "s", "type": "bytes32"}],
  "name": "redeemToken",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [{"name": "", "type": "uint256"}],
  "name": "redeemedTokenIds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "payoutForServiceProvider",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminationTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "closeThroughServiceProvider",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "soonExpiringTokenId",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "terminationStartedAt",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceProvider",
  "outputs": [{"name": "", "type": "address"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "costPerRequest",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [{"name": "newTokenID", "type": "uint256"}],
  "name": "startTokenRedemption",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceConsumerFunds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "serviceProviderFunds",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "soonExpiringTokenIdTimeout",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": true,
  "inputs": [],
  "name": "activeTokenId",
  "outputs": [{"name": "", "type": "uint256"}],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "startCloseThroughServiceConsumer",
  "outputs": [],
  "payable": false,
  "type": "function"
}, {
  "constant": false,
  "inputs": [],
  "name": "depositForServiceConsumer",
  "outputs": [],
  "payable": true,
  "type": "function"
}, {
  "inputs": [{"name": "_serviceProvider", "type": "address"}, {
    "name": "_initialTokenId",
    "type": "uint256"
  }, {"name": "_costPerRequest", "type": "uint256"}, {
    "name": "_terminationTimeout",
    "type": "uint256"
  }, {"name": "_tokenRedemptionTimeout", "type": "uint256"}], "payable": false, "type": "constructor"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "code", "type": "uint8"}],
  "name": "Error",
  "type": "event"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "sum", "type": "uint256"}],
  "name": "Redeemed",
  "type": "event"
}, {
  "anonymous": false,
  "inputs": [{"indexed": false, "name": "sum", "type": "uint256"}],
  "name": "Payout",
  "type": "event"
}];
