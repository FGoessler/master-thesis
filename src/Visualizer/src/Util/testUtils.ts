import { HTMLAttributes, ReactWrapper } from "enzyme";

/// This is a helper function to deal with the bit messy capabilities of enzyme to simulate form input element changes.
export function simulateInput(node: ReactWrapper<HTMLAttributes, {}>, input: string) {
  (node as any).node.value = input;
  node.first().simulate("change");
}

export function noOp() {
  return;
}
