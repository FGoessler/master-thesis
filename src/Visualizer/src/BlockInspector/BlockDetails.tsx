import * as React from "react";
import { Block, Transaction } from "../Util/web3";
import StringFormatter from "../Util/stringFormatter";
import TransactionDetails from "./TransactionDetails";

export default function BlockDetails(props: { block: Block, addressSelectedHandler: (address: string) => void }) {
  return (
    <div>
      <h3>
        {StringFormatter.trimHash(props.block.hash)}
      </h3>
      <table className="detailsTable">
        <tbody>
        <tr>
          <td>Hash</td>
          <td className="hash">{StringFormatter.trimHash(props.block.hash)}</td>
        </tr>
        <tr>
          <td>Number</td>
          <td>{props.block.number}</td>
        </tr>
        <tr>
          <td>Timestamp</td>
          <td>{new Date(props.block.timestamp * 1000).toUTCString()}</td>
        </tr>
        <tr>
          <td>Parent Hash</td>
          <td className="hash">{StringFormatter.trimHash(props.block.parentHash)}</td>
        </tr>
        <tr>
          <td>Transactions</td>
          <td>
            <ul>
              {props.block.transactions.map((t: Transaction) =>
                <TransactionDetails
                  transaction={t}
                  addressSelectedHandler={props.addressSelectedHandler}
                  key={t.hash}
                />
              )}
            </ul>
          </td>
        </tr>
        </tbody>
      </table>
    </div>
  );
}
