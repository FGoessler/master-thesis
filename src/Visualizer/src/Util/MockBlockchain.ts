
import { Block, Eth, Web3, Transaction, TransactionReceipt } from "./web3";
import { BigNumber, BigNumberHelper } from "./bigNumber";

export class MockBlock implements Block {
  number: number = 1;
  hash: string = "0x2a";
  parentHash: string = "0x1a";
  timestamp: number = 100000;
  transactions: (string | Transaction)[] = [];

  constructor(num: number = 1) {
    this.number = num;
  }
}

export class MockTransaction implements Transaction {
  hash: string = "0xffff1a";
  from: string = "0xdd1a";
  to: string  = "0xdd2b";
  value: BigNumber = BigNumberHelper.toBigNumber(0);
  input: string = "";
  transactionReceipt?: TransactionReceipt;
  isContractDeployment: boolean = false;
}

export default class MockWeb3 implements Web3 {

  public eth: Eth;

  constructor() {
    this.eth = new MockEth();
  }
}

export class MockEth implements Eth {

  public transactionReceipts: { [key: string]: TransactionReceipt };
  public blocks: Block[];
  public balances: {};

  constructor() {
    this.transactionReceipts = {
      "0xffff1a": {
        blockHash: "0x2a",
        blockNumber: 1,
        transactionHash: "0xffff1a",
        gasUsed: 100000,
        logs: []
      }
    };

    this.blocks = [
      {
        number: 0,
        hash: "0x1a",
        parentHash: "",
        timestamp: 100000,
        transactions: []
      },
      {
        number: 1,
        hash: "0x2a",
        parentHash: "0x1a",
        timestamp: 100000,
        transactions: [
          new MockTransaction()
        ]
      },
      {
        number: 2,
        hash: "0x3a",
        parentHash: "0x2a",
        timestamp: 100000,
        transactions: []
      }
    ];

    this.balances = {
      "0x1234": BigNumberHelper.toBigNumber("9900000")
    };
  }

  contract(abi: any[]) {
    return {};
  }

  getBlock(hash: number | string, inclTransactions: boolean, cb: (err?: Error, b?: Block) => void): void {
    if (typeof hash === "number") {
      cb(undefined, this.blocks[hash]);
    } else {
      if (hash === "latest") {
        cb(undefined, this.blocks[this.blocks.length - 1]);
      } else {
        const foundBlock = this.blocks.find((b) => { return b.hash === hash; });
        if (foundBlock) {
          cb(undefined, foundBlock);
        } else {
          cb(new Error(`Block ${hash} not found!`), undefined);
        }
      }
    }
  }

  getTransactionReceipt(hash: string, cb: (err?: Error, b?: TransactionReceipt) => void): void {
    if (this.transactionReceipts[hash]) {
      cb(undefined, this.transactionReceipts[hash]);
    } else {
      cb(new Error(`Transaction ${hash} not found!`), undefined);
    }
  }

  getBalance(address: string, cb: ((err?: Error, b?: BigNumber) => void)): void;
  getBalance(address: string, block: number | string, cb: (err?: Error, b?: BigNumber) => void): void;
  getBalance(address: string,
             blockOrCb: number | string | ((err?: Error, b?: BigNumber) => void),
             cb?: ((err?: Error, b?: BigNumber) => void)): void {
    let callback: (err?: Error, b?: BigNumber) => void;
    if (typeof blockOrCb === "function") {
      callback = blockOrCb;
    } else if (cb && typeof cb === "function") {
      callback = cb;
    } else {
      throw new Error("Missing callback!");
    }

    const balance = this.balances[address];
    if (balance) {
      callback(undefined, balance);
    } else {
      callback(new Error(`Address ${address} not found!`), undefined);
    }
  }
}
