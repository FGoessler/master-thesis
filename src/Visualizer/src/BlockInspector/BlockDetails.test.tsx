import * as React from "react";
import * as ReactDOM from "react-dom";
import BlockDetails from "./BlockDetails";
import { MockBlock, MockTransaction } from "../Util/MockBlockchain";
import { noOp } from "../Util/testUtils";

it("shows information about a block loaded from the blocklist service", async () => {
  const mockBlock = new MockBlock(1);
  mockBlock.transactions = [new MockTransaction()];
  const item = document.createElement("div");
  ReactDOM.render(<BlockDetails block={mockBlock} addressSelectedHandler={noOp}/>, item);
  expect(item).toMatchSnapshot();
});
