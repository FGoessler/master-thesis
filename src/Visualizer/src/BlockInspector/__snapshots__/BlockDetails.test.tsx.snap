// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows information about a block loaded from the blocklist service 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x2a
    </h3>
    <table
      class="detailsTable"
    >
      <tbody>
        <tr>
          <td>
            Hash
          </td>
          <td
            class="hash"
          >
            0x2a
          </td>
        </tr>
        <tr>
          <td>
            Number
          </td>
          <td>
            1
          </td>
        </tr>
        <tr>
          <td>
            Timestamp
          </td>
          <td>
            Fri, 02 Jan 1970 03:46:40 GMT
          </td>
        </tr>
        <tr>
          <td>
            Parent Hash
          </td>
          <td
            class="hash"
          >
            0x1a
          </td>
        </tr>
        <tr>
          <td>
            Transactions
          </td>
          <td>
            <ul>
              <li>
                <!-- react-text: 22 -->
                0xffff1a
                <!-- /react-text -->
                <br />
                <!-- react-text: 24 -->
                From:
                <!-- /react-text -->
                <a
                  class="fromAddress hash"
                >
                  0xdd1a
                </a>
                <br />
                <!-- react-text: 27 -->
                To:
                <!-- /react-text -->
                <a
                  class="toAddress hash"
                >
                  0xdd2b
                </a>
                <br />
                <!-- react-text: 30 -->
                Sum: 
                <!-- /react-text -->
                <span
                  class="hash"
                >
                  <!-- react-text: 32 -->
                  0
                  <!-- /react-text -->
                  <!-- react-text: 33 -->
                  WEI
                  <!-- /react-text -->
                </span>
                <br />
                <!-- react-text: 35 -->
                Gas used: 
                <!-- /react-text -->
                <span
                  class="hash"
                >
                  <!-- react-text: 37 -->
                  -
                  <!-- /react-text -->
                  <!-- react-text: 38 -->
                  WEI
                  <!-- /react-text -->
                </span>
                <br />
                <!-- react-text: 40 -->
                Transaction Data:
                <!-- /react-text -->
                <div
                  class="json"
                />
                <!-- react-text: 42 -->
                Logs:
                <!-- /react-text -->
                <div
                  class="json"
                >
                  No logs
                </div>
              </li>
            </ul>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
